sphinx/tutorial/rose/furthertopics/failif-warnif.rst

```diff
@@ -0,0 +1,240 @@
+
+There is also an array shorthand, which we'll demonstrate later.
+
+Note that the ``;`` at the end is optional when we only have one expression
+(it's a delimiter), but it's better style to keep it.
+
+
+Example
+-------
+
+We'll use the example of a rocket launch.
+
+Create a new directory somewhere and run the following command replacing
+``DIRECTORY`` with the path to this new directory::
+
+   rose tutorial fail-if-warn-if DIRECTORY
+   cd DIRECTORY
+
+You will now have a new rose app with a ``rose-app.conf`` that looks like this:
+
+.. code-block:: rose
+
+   [command]
+   default=launch.exe
+
+   [env]
+   ORBITAL_SPEED_MS=1683.0
+
+   [file:rocket_settings.nl]
+   source=namelist:rocket
+
+   [namelist:rocket]
+   battery_levels=80, 60
+   total_weight_kg=4700.0
+   fuelless_weight_kg=2353.0
+   specific_impulse_s=311.0
+
+.. image:: http://upload.wikimedia.org/wikipedia/commons/thumb/2/2a/Apollo16LM.jpg/533px-Apollo16LM.jpg
+   :align: right
+   :alt: Apollo 11 Lunar Module, returning from the surface of the Moon
+   :width: 250px
+
+This app configuration controls the liftoff of a particular rocket -
+in our case, the Lunar Module (Apollo Program spacecraft).
+
+There is also metadata in the ``meta/rose-meta.conf`` file which provides the
+application inputs with descriptions, help text and type information.
+
+Try running rose edit in the app directory. You should be able to navigate
+between the pages and view the help and description for the settings.
+
+
+``fail-if``
+-----------
+
+If the ratio of rocket fuel to total weight is too high, or the efficiency
+of the rocket (specific impulse) is too low, the Lunar Module will never
+make it off the Moon.
+
+We want to be able to flag an error based on a combination of the rocket
+settings and the necessary orbital velocity (``env=ORBITAL_VELOCITY_MS``).
+We need to set some ``fail-if`` metadata on one of these settings - as
+it's evaluated on-demand, it doesn't matter which one we choose.
+
+Open the ``meta/rose-meta.conf`` file in a text editor.
+
+Add the following line to the metadata section
+``[namelist:rocket=total_weight_kg]``:
+
+.. code-block:: rose
+
+   fail-if=this < namelist:rocket=fuelless_weight_kg * 2.7183**(env=ORBITAL_SPEED_MS / (9.8 * namelist:rocket=specific_impulse_s));
+
+This states the relationship between these settings (a rearrangement
+of the `Tsiolkovsky rocket equation`_). The rocket must have a
+sufficient ratio of fuel to rocket mass, with a sufficiently fast
+exhaust velocity (``=9.8 * namelist:rocket=specific_impulse_s``)
+to get to the orbital speed ``env=ORBITAL_SPEED_MS``.
+
+Save the metadata file and then reload the config editor metadata.
+You now need to ask Rose to evaluate the ``fail-if`` condition, as
+it's an on-demand process.
+
+Either press the toolbar button :guilabel:`Check fail-if ...` or click
+the menu :menuselection:`Metadata --> Check fail-if, warn-if`.
+
+Hopefully, this should not flag any errors, as these are the Apollo
```

Comment:
It took me a while to spot the config editor output this refers to (though I have little experience with config-edit so it could just be me). Perhaps mention it displays messages at the bottom of the GUI (as opposed to a pop-up window or printing to the shell etc.)